durable-storage/src/merkle_layer/node.rs

```diff
@@ -114,18 +173,49 @@ impl MavlNode {
     fn invalidate_hash(&mut self) {
         self.hash = OnceLock::new();
     }
+
+    pub(super) fn decode(
+        hash_representation: MavlNodeHashRepresentation,
+        commited_hash: blake3::Hash,
+        left: Wrapper,
+        right: Wrapper,
+    ) -> Self {
+        Self {
+            key: Key::new(hash_representation.key).expect("They keys are alright"),
+            data: BytesMut::from(hash_representation.data),
+            left,
+            right,
+            hash: OnceLock::from(commited_hash),
+            balance_factor: hash_representation.balance_factor,
+        }
+    }
+
+    fn resolve_children<Resolver: MavlNodeResolver<NodeWrapper = Wrapper>>(
+        &self,
+        node_resolver: &Resolver,
+    ) {
+        node_resolver.resolve(&self.left);
+        node_resolver.resolve(&self.right);
+    }
 }
 
 /// Delete the value of the node with a given key. If the key does not exist, do nothing.
 ///
 /// Returns true if the subtree has shrank in size.
-pub(super) fn delete(root: &mut Option<Arc<MavlNode>>, key: &Key) -> bool {
+pub(super) fn delete<
+    Wrapper: MavlNodeWrapper,
+    Resolver: MavlNodeResolver<NodeWrapper = Wrapper>,
+>(
+    root: &mut Option<Arc<MavlNode<Wrapper>>>,
+    key: &Key,
+    node_resolver: &Resolver,
+) -> bool {
```

Comment:
Something is fishy with these methods (same for `get`, etc). There are too many type variables, and something strikes me as quite odd that there is an `Option<Arc<_>>` in the mix, but the traits introduced in this PR allow you to actually not be `Option<Arc<_>>`. So it seems this just happens to work as long as all node wrapper types somehow satisfy the `Option<Arc<_>>` requirement. 

```suggestion
pub(super) fn delete<
    Resolver: MavlNodeResolver,
>(
    root: &mut Resolver::NodeWrapper,
    key: &Key,
    node_resolver: &Resolver,
) -> bool {
```